File: 1/1-8.ts
function improve(guess: number, x: number): number {
    return (x / (guess * guess) + 2 * guess) / 3;
}

function goodEnough(guess: number, x: number): boolean {
    return Math.abs(guess * guess * guess - x) < 0.001;
}

function cubeRootIter(guess: number, x: number): number {
    if (goodEnough(guess, x)) {
        return guess;
    } else {
        return cubeRootIter(improve(guess, x), x);
    }
}

function cubeRoot(x: number): number {
    return cubeRootIter(1.0, x);
}

console.log(cubeRoot(27.0));

File: 1/1-8.rs
fn improve(guess: f64, x: f64) -> f64 {
    (x / (guess * guess) + 2.0 * guess) / 3.0
}

fn good_enough(guess: g64, x: f64) -> bool {
    (guess * guess * guess - x).abs() < 0.001
}

fn cube_root_iter(guess: f64, x: f64) -> f64 {
    if good_enough(guess, x) {
        guess
    } else {
        cube_root_iter(improve(guess, x), x)
    }
}

fn cube_root(x: f64) -> f64 {
    cube_root_iter(1.0, x)
}

fn main() {
    println!("{}", cube_root(27.0));
}